rnd/autogpt_server/autogpt_server/server/server.py

```diff
@@ -178,20 +182,63 @@ def run_service(self):
             methods=["POST"],
         )
 
-        app.add_exception_handler(500, self.handle_internal_error)  # type: ignore
+        app.add_exception_handler(500, self.handle_internal_error)
 
-        app.mount(
-            path="/frontend",
-            app=StaticFiles(directory=get_frontend_path(), html=True),
-            name="example_files",
-        )
+        class SPAStaticFiles(StaticFiles):
+            async def get_response(
+                self, path: str, scope: MutableMapping[str, Any]
+            ) -> Response:
+                try:
+                    return await super().get_response(path, scope)
+                except (HTTPException, StarletteHTTPException) as ex:
+                    if ex.status_code == 404:
+                        return await super().get_response("build.html", scope)
+                    else:
+                        raise ex
 
         app.include_router(router)
 
+        #! WARNING: THE ORDER OF THE REST OF THESE FUNCTIONS IS IMPORTANT
+        #! DO NOT CHANGE THE ORDER OF THESE FUNCTIONS
+        #! DO NOT ADD ANY FUNCTIONS BETWEEN THESE FUNCTIONS
+        #! DO NOT REMOVE ANY OF THESE FUNCTIONS
+        #! YOU WILL BREAK THE FRONTEND PACKAGED WITH THE SERVER
+
+        if os.path.exists(get_frontend_path() / "_next"):
+            app.mount(
+                path="/_next",
+                app=SPAStaticFiles(directory=get_frontend_path() / "_next", html=True),
+                name="frontend",
+            )
+
         @app.websocket("/ws")
         async def websocket_endpoint(websocket: WebSocket):  # type: ignore
             await self.websocket_router(websocket)
 
+        @app.get("/{extras_file:str}", response_class=HTMLResponse)
+        async def catch_others(extras_file: str):
+            # Do not allow directory traversal
+            if ".." in extras_file:
+                raise HTTPException(status_code=400, detail="Invalid file path")
+            if extras_file.count(".") > 1:
+                raise HTTPException(status_code=400, detail="Invalid file path")
+            # check if the file is directly in the frontend path folder to further prevent directory traversal against the allowed files
+            files = os.listdir(get_frontend_path())
+            # detect the file type and return the appropriate response
+            if extras_file in files:
+                mime = mimetypes.guess_type(extras_file)[0]
+                if mime is not None and mime.startswith("image"):
+                    return FileResponse(get_frontend_path() / extras_file)
+                else:
+                    with open(get_frontend_path() / extras_file) as file:
+                        return str(file.read())
+            raise HTTPException(status_code=400, detail="Invalid file path")
+
+        @app.get("/", response_class=HTMLResponse)
+        async def catch_root():
+            with open(get_frontend_path() / "build.html") as file:
+                return str(file.read())
```

Comment:
This shouldn't be necessary once `next.config.js` is merged into `next.config.mjs`, which contains a route for `/` -> `/build`.

Reply:
This is for the api side to do the same when serving it

Reply:
That's understood. Still, probably not necessary as Next.js will generate a route for `/` that handles the redirect. Having that logic in two different places - let alone two different codebases - is less than ideal. I also doubt that `/build` will be the default home page for much longer.